### PR TITLE
Rework completely how command line options are handled

### PR DESCRIPTION
**Pull request by xclaesse**
The code was a mess and had many small bugs and inconsistencies between mesonmain and mconf. This PR centralize into coredata all command line parsing. See individual commits for details.

### CODE REVIEW

**Comment by Akaricchi** [comment]
Nice. Does this also address #3456? Seems like the perfect opportunity.

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2182,6 +2182,35 @@ def parse_default_options(self, default_options):
                 newoptions = [defopt] + self.environment.cmd_line_options.projectoptions
                 self.environment.cmd_line_options.projectoptions = newoptions
 
+    def set_backend(self):
+        if self.backend is not None:
+            return
```
Comment (nirbheek) [inline mesonbuild/interpreter.py]: Why is this needed? A comment might be in order.

Comment (xclaesse) [reply mesonbuild/interpreter.py]: When parsing subproject the backend is already set.

**Comment by xclaesse** [comment]
@nirbheek added a comment in set_backend(). Also pushed more commits to use argparse instead of manual parsing args as much as possible, and add reconfigure/wipe features.

**Comment by xclaesse** [comment]
oh, commit order was completely wrong in github because I did a log of `git rebase -i` to cleanup the history and github present in chronological order instead of topological order. `git rebase --ignore-date master` fixed the ordering.

It is important to review this branch in the right order, commits are really one step at a time.

**Comment by xclaesse** [comment]
Hm, even with the rebase, the order seems wrong in the "conversation" tab, but seems right in the "Commits" tab.

https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -753,7 +753,27 @@ def log_var(var, val):
         return preproc_flags, compile_flags, link_flags
 
     def get_options(self):
-        return {} # build afresh every time
+        opts = {} # build afresh every time
+
+        # Take default values from env variables.
```
Comment (nirbheek) [inline mesonbuild/compilers/compilers.py]: How does this change reconfiguration? Earlier we would always rebuild this from scratch, but now it's merged with compiler options and those are restored from coredata.

File: mesonbuild/coredata.py [hunk 3]
```diff
@@ -332,6 +327,15 @@ def validate_option_value(self, option_name, override_value):
                 return opt.validate_value(override_value)
         raise MesonException('Tried to validate unknown option %s.' % option_name)
 
+    def external_args(self, lang):
+        return self.compiler_options[lang + '_args'].value
+
+    def external_link_args(self, lang):
+        return self.compiler_options[lang + '_link_args'].value
+
+    def external_preprocess_args(self, lang):
```
Comment (nirbheek) [inline mesonbuild/coredata.py]: Shouldn't this be `get_external_*()`? That's how class methods usually are.

File: mesonbuild/compilers/compilers.py (hunk 2 above)
Comment (nirbheek) [reply mesonbuild/compilers/compilers.py]: Also, does this mean that `CFLAGS` from the environment are also added for the cross-compiler now? That must not happen.

File: mesonbuild/coredata.py [hunk 4]
```diff
@@ -141,34 +141,25 @@ class UserArrayOption(UserOption):
...
-        # string, but for defining options in meson_options.txt the format
-        # should match that of a combo
-        if not user_input:
-            if isinstance(value, str):
-                if not value.startswith('['):
-                    raise MesonException('Valuestring does not define an array: ' + value)
-                newvalue = ast.literal_eval(value)
-            else:
-                newvalue = value
-        else:
```
Comment (nirbheek) [inline mesonbuild/coredata.py]: After reading the commit list, it's not clear to me what the replacement is for parsing of options in cross files.

File: run_unittests.py [hunk 5]
```diff
@@ -2163,6 +2163,33 @@ def test_command_line(self):
         self.assertEqual(obj.compiler_options['c_args'].value, ['foo bar', 'one', 'two'])
         self.wipe()
```
Comment (nirbheek) [inline run_unittests.py]: Please edit the commit message to say "Closes #3456" so it is closed automatically.

File: mesonbuild/environment.py [hunk 6]
```diff
@@ -274,7 +274,7 @@ def __init__(self, source_dir, build_dir, options):
             self.cross_info = CrossBuildInfo(self.coredata.cross_file)
         else:
             self.cross_info = None
-        self.cmd_line_options = options.cmd_line_options
+        self.cmd_line_options = copy.copy(options.cmd_line_options)
```
Comment (nirbheek) [inline mesonbuild/environment.py]: If you don't need a deep copy, this can just be `options.cmd_line_options.copy()`?

File: mesonbuild/mesonmain.py [hunk 7]
```diff
@@ -40,6 +40,8 @@ def create_parser():
                    help='Special wrap mode to use')
     p.add_argument('--profile-self', action='store_true', dest='profile',
                    help=argparse.SUPPRESS)
+    p.add_argument('--reconfigure', action='store_true',
+                   help='Reconfigure the project using the same options.')
```
Comment (nirbheek) [inline mesonbuild/mesonmain.py]: This is a new feature, so it needs @jpakkane's approval.

File: mesonbuild/mlog.py [hunk 8]
```diff
@@ -50,9 +50,12 @@ def initialize(logdir):
 def shutdown():
     global log_file
     if log_file is not None:
+        path = log_file.name
         exception_around_goer = log_file
         log_file = None
         exception_around_goer.close()
+        return path
+    return None
```
Comment (nirbheek) [inline mesonbuild/mlog.py]: Why?

File: mesonbuild/coredata.py [hunk 9]
```diff
@@ -194,7 +195,7 @@ def __init__(self, options):
         self.compiler_options = {}
         self.base_options = {}
         self.cross_file = self.__load_cross_file(options.cross_file)
-        self.wrap_mode = options.wrap_mode
+        self.wrap_mode = options.wrap_mode if options.wrap_mode is not None else WrapMode.default
```
Comment (nirbheek) [inline mesonbuild/coredata.py]: Why is this needed? How is `None` different from `default`?

File: mesonbuild/mesonmain.py [hunk 10]
```diff
@@ -33,7 +33,7 @@ def add_setup_arguments(p):
     p.add_argument('-v', '--version', action='version',
                    version=coredata.version)
     # See the mesonlib.WrapMode enum for documentation
-    p.add_argument('--wrap-mode', default=WrapMode.default,
+    p.add_argument('--wrap-mode', default=None,
```
Comment (nirbheek) [inline mesonbuild/mesonmain.py]: This also changes the `--help` output, so if you need to differentiate between `None` and `default` this needs to be done in a different way.

File: mesonbuild/wrap/wraptool.py [hunk 11]
```diff
@@ -207,9 +206,4 @@ def status(options):
 
 def run(options):
     options.wrap_func(options)
-
-if __name__ == '__main__':
-    parser = argparse.ArgumentParser(prog='wraptool')
-    add_arguments(parser)
-    options = parser.parse_args(sys.argv[1:])
-    run(options)
+    return 0
```
Comment (nirbheek) [inline mesonbuild/wrap/wraptool.py]: Why do you add this in the commit porting wraptool to argparse and then remove it here?

File: mesonbuild/compilers/cpp.py [hunk 12]
```diff
@@ -148,6 +151,8 @@ def __init__(self, exelist, version, gcc_type, is_cross, exe_wrapper=None, defin
 
     # It does not support c++/gnu++ 17 and 1z, but still does support 0x, 1y, and gnu++98.
     def get_options(self):
+        # FIXME: Is it intentional to pull C options here? With multi-inheritage
+        # it is unclear if this pulls GnuCPPCompiler or ElbrusCompiler options.
         opts = super().get_options()
```
[the remark this answers is not in the document]
Comment (xclaesse) [reply mesonbuild/compilers/cpp.py]: ok, done.

File: mesonbuild/compilers/compilers.py [hunk 13]
```diff
@@ -57,6 +57,15 @@
...
 soregex = re.compile(r'.*\.so(\.[0-9]+)?(\.[0-9]+)?(\.[0-9]+)?$')
 
+# Environment variables that each lang uses.
+cflags_mapping = {'c': 'CFLAGS',
+                  'cpp': 'CXXFLAGS',
+                  'objc': 'OBJCFLAGS',
+                  'objcpp': 'OBJCXXFLAGS',
+                  'fortran': 'FFLAGS',
+                  'd': 'DFLAGS',
+                  'vala': 'VALAFLAGS'}
```
[the remark this answers is not in the document]
Comment (xclaesse) [reply mesonbuild/compilers/compilers.py]: ok, done.

File: mesonbuild/compilers/compilers.py (hunk 2 above)
Comment (xclaesse) [reply mesonbuild/compilers/compilers.py]: > How does this change reconfiguration? Earlier we would always rebuild this from scratch, but now it's merged with compiler options and those are restored from coredata.

I'm not sure to understand what you mean. When you run `ninja reconfigure` it pulls new flags from env? I don't think that was working like that previously, since flags were stored in coredata at first configuration. 

Before my patch the code was doing:
```
            if lang not in self.coredata.external_args:
                self.coredata.external_args[lang] = compile_args
```
So on reconfigure, lang is already in `self.coredata.external_args` and it won't set the value from env.

> Also, does this mean that CFLAGS from the environment are also added for the cross-compiler now? That must not happen.

No, interpreter.py, detect_compilers(), it calls get_options() only on the native compiler. It is the backend who decide if it takes those values from coredata, see for example ninja backend does:

```
            if not target.is_cross:
                # Link args added from the env: LDFLAGS. We want these to
                # override all the defaults but not the per-target link args.
                commands += self.environment.coredata.external_link_args(linker.get_language())
```

File: mesonbuild/coredata.py (hunk 3 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: ok, done.

File: mesonbuild/environment.py (hunk 6 above)
Comment (xclaesse) [reply mesonbuild/environment.py]: Ok, didn't know dict had a copy() method. Done.

File: mesonbuild/mesonmain.py (hunk 7 above)
Comment (xclaesse) [reply mesonbuild/mesonmain.py]: Right, I could split my PR before that commit so the rest could still go. Is there a way in github to tell a PR depends on another PR ?

File: mesonbuild/mlog.py (hunk 8 above)
Comment (xclaesse) [reply mesonbuild/mlog.py]: Because commit "Use a single ArgumentParser for all subcommands" uses it.

It is nicer to get the log file from mlog.shutdown() instead of doing `logfile = os.path.join(app.build_dir, environment.Environment.log_dir, mlog.log_fname)` because it's mlog who decide the path where it store its logs, not mesonmain.

File: mesonbuild/coredata.py (hunk 9 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: Because we need to know if the user explicitly passed --wrap-mode=default or if he omitted the option from the command line. This is important in the next commit "mesonmain: Add --wipe argument".

Let's say user configure with `--wrap-mode=forcefallback`, the cmd_line.txt file will contain `wrap_mode=forcefallback`. Now the user wipe his build dir with `meson setup --wipe` we'll have options.wrap_mode=default, does it mean user ran the command `meson setup --wipe --wrap-mode=default` to explicitly override that value from cmd_line.txt, or does it mean he didn't pass any value and thus we should still used `wrap_mode=forcefallback`?

To fix that ambiguity, when user does not set --wrap-mode in the command line, options.wrap_mode is None.

File: mesonbuild/mesonmain.py (hunk 10 above)
Comment (xclaesse) [reply mesonbuild/mesonmain.py]: No the default is not printed by --help. That's actually a good point, for builtin options we manually add '(default: %s)' to the help argument, but we don't do that for other options. But for --wrap-mode the default is pretty obvious anyway, since one of the choices is named "default".

File: mesonbuild/wrap/wraptool.py (hunk 11 above)
Comment (xclaesse) [reply mesonbuild/wrap/wraptool.py]: Because wraptool had a __main__ that parse manually command line, so I just ported it as-is using argparse. It's only later I realized that __main__ is legacy that doesn't even work anymore, so I removed it in a few places.

**Comment by xclaesse** [comment]
Thanks for your review, I know it's a lot of work to review such monster PR, but it was also a lot of work to write it ;-) Trying to fix just a small thing, then discover ugly stuff all over the place that needs cleanup... That was a trap...

> This commit list is rather difficult to review. In the future you might want to split such work across several independent PRs.

It cannot really be splitted into different PR because each commit is based on the previous one. I tried really hard to split in self contained commits to make the review easier. github is making its best at making reviewer's life the hardest possible by reordering my commits in chronological order which makes no sense at all when you try to make nice commits using git rebase -i all the time. It really needs to be reviewed in topological order, better checkout the branch locally and see `git log` ;-)

> These changes will almost certainly cause regressions, especially because we don't have comprehensive tests for all the cases of command-line parsing.

I wrote a few more tests, and ran many tests myself. But indeed regressions are unavoidable here I'm afraid. IMO it's worth it. Hopefully it fix more bugs that it created. But at least I'm sure if we find bugs they'll be easier to fix after that massive cleanup.

**Comment by xclaesse** [comment]
@nirbheek if you're happy with the first X commits, let me know and I'll create a separate PR with that subset.

File: mesonbuild/coredata.py (hunk 4 above)
Comment (jpakkane) [reply mesonbuild/coredata.py]: Yes. In Meson files we always use explicit types as much possible. String splitting is only for data that comes from the outside because there is really no way to do it otherwise.

Comment (xclaesse) [reply mesonbuild/coredata.py]: ok, kept a small cleanup patch there, but not changing the behaviour.

File: mesonbuild/interpreter.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/interpreter.py]: This commit fix 3 things, maybe commit message wasn't clear:

1) `project(default_options : 'backend=msvc')` silently has no effect because the backend was already decided at that point. I agree it's corner case but I find it weird that it's the only option not behaving like all others. At the very least we should warn. But since I made the change already, better keep it IMO.
2) `project(default_options : 'backend_max_links=42')` silently has no effect, no backend-specific option gets parsed there.
3) `meson -Dbackend_max_links=42` silently has no effect. Actually the only way to set that option is with `meson configure`.

I think consistency is really important, for ALL options we must respect:
cmdline > subproject() default > project() default > meson_options.txt/builtin default

**Comment by nirbheek** [comment]
Could you move the new flags `--wipe` and `--reconfigure` out of this PR? Once this PR is bugfix-only I can do another review of this and push it.

**Comment by xclaesse** [comment]
I created PR  #3612 that does not include new features.

**Comment by xclaesse** [comment]
Rebased now that the first batch of commits got merged.

**Comment by jpakkane** [comment]
Is this still relevant? If yes, it ⚔️ conflicts.

**Comment by xclaesse** [comment]
I've splitter this PR in pieces, let's close this one. Currently at step 3 : pr #4077
